### DISCUSSION TITLE
Detection of PHP version if multiple are supported + composer lockfiles

### DISCUSSION THREAD

**Discussion by herndlm** [renovatebot/renovate#18840] [Request Help]
Hi,

I got another one related to renovate dynamically detecting the PHP version it will use for composer runs. This was implemented recently for `binarySource=install` too and it works like the rest in Renovate I suppose. E.g. if we have a composer PHP platform requirement like `^7.4 || ^8.0`, Renovate detects that it can use the latest PHP 8.1, which makes sense, since we explicitly state that we support it.

But then composer comes into play. When doing e.g. `composer require foo -W`, composer updates the dependencies of foo. If PHP 8.1 is used, composer will temporarily set the PHP constraint to the currently used version and potentially pull in packages that are not supported by PHP 7.4 leading to an invalid lockfile. I thought that behaviour is weird and asked via https://github.com/composer/composer/discussions/11164 but this seems to be completely normal. A workaround is to set a platform constraint which also makes things work in Renovate again accordingly, but I wonder - is there another, smarter, way to improve this situation that doesn't involve using a platform constraint? Basically, I'd need something like Renovate to  select the lowest minor but highest major + patch or so, but most likely I oversimplify it, since the SemVer constraints are more complicated..

**Comment by PhilipAbed** [comment]
so you basically need to update only major and patch versions, and no minor updates?
but then when you move to a new version, renovate automatically takes the highest available version..

how about you create a package rule :
https://docs.renovatebot.com/configuration-options/#allowedversions

can i assume the lowest minor is always 0? you can add a rule that will allow only major/patch changes but keeps minor at 0 all the time like this
```
  "packageRules": [
    {
       "updateTypes": ["minor", "pin"],
      "allowedVersions": "/^[0-9]+\\.0+\\.[0-9]+(\\.[0-9]+)?$/"
    }
  ]
```
you can also match with   "matchPackageNames": ["pkg name"], if there are specific packages you want
you can also cancel the minor updates using another package rule:
```
      "matchUpdateTypes": ["minor"],
      "enabled": false
```

hope this helps

**Reply by rarkins** [reply]
No, that's not it. His question is about which version of PHP is selected by Renovate.
